### ISSUE TITLE
huge amount of memory usage in load_model (tf 2.0)

### ISSUE THREAD

**Issue by jessica-santos** [tensorflow/tensorflow#34107]
Hi, I retrained a Inception + DenseLayer(512) using keras in tf 2.0.

When I load this model with tf 1.14, it uses something around 4G of GPU memory. But, when I'm doing the same with tf 2.0 it uses more than 25Gb of memory!!
This is correctly? That is some other config to load the model now?

I'm using an ec2 with a tesla v100 and cuda 10.1 

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary):
- TensorFlow version (use command below): 1.14 and 2.0 (different environments)
- Python version: 3
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version: 10.1
- GPU model and memory: Tesla V100

**Comment by ravikyram** [comment]
@jessica-santos 

Request you to share minimal standalone  code to reproduce the issue in our environment,it helps us localizing the issue faster. Thanks!

**Comment by jessica-santos** [comment]
Sure, this is the code I run to build and train my model:

`def build_model(classe=3):
    """
    :return: builded model
    """
    base_model = InceptionResNetV2(weights='imagenet', include_top=False, input_shape=(IMAGE_SIZE, IMAGE_SIZE, 3))
    last = base_model.output

    x = GlobalAveragePooling2D()(last)

    x = Dense(512, activation='relu')(x)

    x = Dropout(.3)(x)

    preds = Dense(classe, activation='sigmoid')(x)
    model = Model(base_model.input, preds)

    return model


def compile_model(model, learning_rate=1e-4):
    """
    :param model: builded model
    :return: compiled model
    """

    optimizer = Adam(lr=learning_rate)
    model.compile(loss='binary_crossentropy', optimizer=optimizer, metrics=['acc'])
    return model`

`core_idg = ImageDataGenerator(brightness_range=[0.3, 0.5],
                              rotation_range=35, 
                              fill_mode='nearest',
                              zoom_range=0.125,
                              rescale = 1./255)

train_generator = core_idg.flow_from_dataframe(train_df, directory=None, x_col='Path', 
                                    y_col=classes_patologias, class_mode='other',
                                    target_size=(IMAGE_SIZE, IMAGE_SIZE), batch_size=BATCH_SIZE)

val_generator = core_idg.flow_from_dataframe(valid_df, directory=None, x_col='Path', 
                                    y_col=classes_patologias, class_mode='other',
                                    target_size=(IMAGE_SIZE, IMAGE_SIZE), batch_size=BATCH_SIZE)`


`model = build_model(classe=len(classes_patologias))
model = compile_model(model)`

`checkpoint = ModelCheckpoint(
    filepath='model.h5',
    monitor='val_loss',
    verbose=1,
    save_best_only=True,
    mode='min',
    period=1
)
`

`model.fit_generator(
    generator=train_generator,
    steps_per_epoch=len(train_df) // BATCH_SIZE,
    epochs=50,
    validation_data=val_generator,
    validation_steps=len(valid_df) // BATCH_SIZE,
    class_weight=sample_weights,
    verbose=1,
    callbacks=[checkpoint, reduce_on_plateau, early_stop],
    workers=4,
    max_queue_size=16,
    use_multiprocessing=False
)`

This is the simple code to load the model:

`from tensorflow.keras.models import load_model
model = load_model('model.h5')`

**Comment by ravikyram** [comment]
@jessica-santos 

Looks like code is incomplete. Request you to provide complete code snippet to reproduce the issue in our environment. Thanks!

**Comment by jessica-santos** [comment]
The only thing missing was those two callbacks:
```
early_stop = EarlyStopping(monitor='val_loss',
                               min_delta=0.0001,
                               patience=5,
                               mode='min',
                               verbose=1)

reduce_on_plateau = ReduceLROnPlateau(
    monitor='val_loss',
    factor=0.1,
    patience=2,
    verbose=1,
    mode='min',
    min_delta=0.01,
    cooldown=0,
    min_lr=0
) 
```

**Comment by ravikyram** [comment]
Looks like some undefined variables like `train_df, valid_df` are there in the code. Request you to provide the same . Thanks!

**Comment by jessica-santos** [comment]
As the name suggests, those are dataframes with train and valid data respectively. The columns are: `'Path', 'class_1','class_2','class_3'`.  `Path` contains the full path to the image and the classes variables contain 1 or 0 for each class.

The variable `classes_patologias` is the list `['class_1','class_2','class_3']`.

Also:
```
IMAGE_SIZE = 384
BATCH_SIZE = 16
```

I cannot share the data, unfortunately.

**Comment by ravikyram** [comment]
@jessica-santos 

Is it possible to share dummy data so we can try to reproduce the issue in our environment. It helps us in localizing the issue faster. Thanks!

**Comment by jessica-santos** [comment]
We use medical images, you can get a free dataset from Kaggle, like this one: https://www.kaggle.com/paultimothymooney/chest-xray-pneumonia

**Comment by rjafarau** [comment]
Any updates?

**Comment by jvishnuvardhan** [comment]
@rjafarau Do you have any simple standalone code to reproduce the issue? Standalone codes results in faster resolution. Thanks!

**Comment by rjafarau** [comment]
@jvishnuvardhan sure

**System information**
Official Tensorflow docker image tensorflow/tensorflow:2.0.0-gpu-py3-jupyter
GPU model and memory: GeForce GTX 1070 Ti 8Gb

```python
import numpy as np

from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import GlobalAveragePooling2D, Dense
from tensorflow.keras.applications import DenseNet121
from tensorflow.keras.callbacks import ModelCheckpoint, EarlyStopping
from tensorflow.keras.models import load_model

N, H, W, C = 9000, 224, 224, 3

example_model = Sequential([
    DenseNet121(input_shape=(H, W, C),
                include_top=False,
                weights='imagenet'),
    GlobalAveragePooling2D(),
    Dense(1, activation='sigmoid')
])

example_model.compile(optimizer='adam',
                      loss='binary_crossentropy',
                      metrics=['accuracy'])

checkpoint_cb = ModelCheckpoint(filepath='example_model.h5',
                                save_best_only=True)
early_stopping_cb = EarlyStopping(patience=5,
                                  restore_best_weights=True)

np.random.seed(42)
history = example_model.fit(
    x=np.random.rand(N, H, W, C),
    y=np.random.randint(2, size=N),
    validation_split=0.2,
    epochs=10,
    batch_size=32,
    callbacks=[checkpoint_cb, early_stopping_cb]
)

# here is a problem
loaded_example_model = load_model('example_model.h5')
```

**Comment by jvishnuvardhan** [comment]
@rjafarau Your files size is very very big. You need to use data pipelines to optimize the memory usage. Thanks!

**Comment by jessica-santos** [comment]
@jvishnuvardhan my original issue is not during the training, is when I load the trained model to do the predict. Do you know if there was any progress on that?

**Comment by jvishnuvardhan** [comment]
@jessica-santos Can you please provide a simple standalone code to reproduce the issue? Without standalone code it is difficult to make any progress. You could take any public data and create a standalone code to reproduce the issue. Thanks!

**Comment by rjafarau** [comment]
@jvishnuvardhan Please make sure that you understand me and @jessica-santos correctly. The problem is not in the training process. I trained a model and saved it. After that, I have an 82Mb hdf5 file. I expect that loading a model from it won't take a lot of RAM. But it doesn't. In my case, it takes over 20Gb.
In my original code, I use ImageDataGenerator with default batch_size. So there is no problem in training.
You have all that you need (standalone code that reproduces the issue).

**Comment by jvishnuvardhan** [comment]
@rjafarau I understand your problem. As I don't have your `82MB hdf5` file, I ran training to save the model with weights. However, data size was too big (I didn't use ImageDatagen) so the program was crashing. I reduced the data size and ran the model, saved (`81.42MB`) and loaded the model without any issue. Please let me know if I am missing anything here. Thanks!

**Comment by rjafarau** [comment]
@jvishnuvardhan How much RAM do you have?

**Comment by jvishnuvardhan** [comment]
@rjafarau I had 12.72 GB ram available but it was not using much. Please check it yourself on the colab (top right corner). Thanks!

**Comment by jvishnuvardhan** [comment]
Closing due to lack of recent activity. Please update the issue when new information becomes available, and we will reopen the issue. Thanks!

**Comment by grkashani** [comment]
I have the same problem , anyone find a issue

**Comment by grkashani** [comment]
I have the same problem , anyone find the issue ?

**Comment by grkashani** [comment]
> @jvishnuvardhan Please make sure that you understand me and @jessica-santos correctly. The problem is not in the training process. I trained a model and saved it. After that, I have an 82Mb hdf5 file. I expect that loading a model from it won't take a lot of RAM. But it doesn't. In my case, it takes over 20Gb.
> In my original code, I use ImageDataGenerator with default batch_size. So there is no problem in training.
> You have all that you need (standalone code that reproduces the issue).

Hi , If you find the problem please let me know 

thx

**Comment by mihaimaruseac** [comment]
Please open a new issue, fill in issue template

**Comment by rjafarau** [comment]
@grkashani In my case, a simple update to the newer version helped me (from 2.0 to 2.1 as I remember). Good luck!

**Comment by raman123456789** [comment]
I have the same issue can anyone help me.

**Comment by johnnync13** [comment]
here is the solution:
https://www.tensorflow.org/api_docs/python/tf/config/experimental/set_memory_growth
